File: DebugTools/C/Prof_eg.c
#include <stdio.h>
#include <math.h>
#include <time.h>
#include <stdlib.h>

#define size 100000

void fill_array(float* array, int array_size);
void show_array(float* array, int array_size);
void solve_array(float* array, int array_size);

void add_to_element(float* element, float increment);
void divide_element(float* element, float divisor);
void multiply_element(float* element, float multiplicand);


int main(int argc, char** argv){

  float * array;
  int i;
  int n_iter = 10000;
  array = malloc(size*sizeof(int));

  fill_array(array, size);

  for(i = 0; i < n_iter; i++){
    printf("%d\n", i);
    solve_array(array, size);
  }

  free(array);
  return 0;
}



void fill_array(float* array, int array_size){
/* Fill passed in array with random values. Array must have been allocated*/

  int i, j=0;
  srand(time(NULL));

  for(i = 0; i < array_size; i++){
    array[i] = (float)(rand()%10000) /30.0;
    if(array[i] > 100.0) j++;
  }
  printf("%d", j);
}

void show_array(float* array, int array_size){
/* Print array to terminal */

  int i;
  for(i = 0; i < array_size; i++){
    printf("%6.3f\t", array[i]);
  }

}

void output_array(float* array, int array_size, char* filename){
/* Print array to file of name */

  int i;
  FILE *fileptr;
  fileptr = fopen(filename, "w");

  for(i = 0; i < array_size; i++){
    fprintf(fileptr, "%6.3f\t", array[i]);
  }

  fclose(fileptr);

}


void solve_array(float* array, int array_size){
/* "Solve" the array. Here we just do some test operations, rather than
a useful algorithm
*/

  int i;
  float inc, div, recip_div;
  inc = 3.0;
  div = 2.0;
  recip_div = 1.0/div;
  for(i = 0; i < array_size; i++){

    add_to_element(array + i, inc);
    if(array[i] > 100.0) divide_element(array + i, div);
    /*if(array[i] > 100.0) multiply_element(array + i, recip_div);*/

  }

}

void add_to_element(float* element, float increment){

  *element += increment;
}

void divide_element(float* element, float divisor){

  *element /= divisor;

}

void multiply_element(float* element, float multiplicand){

  *element *= multiplicand;

}

File: DebugTools/Python/Prof_eg.py
from __future__ import print_function

def fill_array(array):
# Fill passed in array with random values. Array must have been allocated
  import random
  import datetime

  random.seed(datetime.datetime.now())
  array_size = len(array)
  j=0

  for i in range(array_size):
    array[i] = random.random()*10000.0/30.0
    if (array[i] > 100.0): j+=1

  print(j)

def show_array(array):
# Print array to terminal
  array_size = len(array)
  for i in range(array_size):
    print("{}\t".format(array[i], end=""))
  print("")

def ouput_array(array,filename):
# Print array to file of name
  array_size = len(array)
  fh = open(filename,"w")

  for i in range(array_size):
    fh.write("{}\t".format(array[i], end=""))

  fh.close()

def solve_array(array):
# "Solve" the array. Here we just do some test operations, rather than
#a useful algorithm

  array_size = len(array)
  inc = 3.0
  div = 2.0
  recip_div = 1.0/div

  for i in range(array_size):
    if (array > 100.0): array[i] = divide_element(array[i], div)
    #if (array > 100.0): array[i] = multiply_element(array[i], recip_div)

def add_to_element(value, increment):
  return value+increment

def divide_element(value, divisor):
  return value/divisor

def multiply_element(value, multiplicand):
  return value * multiplicand

def main(size = 10000,n_iter=1000):

  array = [0]*size

  fill_array(array)

  for i in range(n_iter):
    print(i)
    solve_array(array)
